text/0000-expand-impl-trait.md

```diff
@@ -0,0 +1,853 @@
+
+One important question is: will people find it easier to understand and use
+`impl Trait`, or something like `some Trait` and `any Trait`? Having an explicit
+split may make it easier to understand what's going on. But on the other hand,
+it's a somewhat complicated distinction to make, and while you usually know
+*intuitively* what you want, being forced to spell it out by choosing the
+correct choice of `some` or `any` seems like an unnecessary burden, especially
+if the choice is almost always dictated by the position.
+
+Pedagogically, if we have an explicit syntax, we retain the option of
+explaining what's going on with `impl Trait` by "desugaring" it into that
+syntax. From that standpoint, `impl Trait` is meant purely for ergonomics, which
+means
+[not just what you type, but also what you have to remember](https://blog.rust-lang.org/2017/03/02/lang-ergonomics.html). Having
+`impl Trait` "just do the right thing" seems pretty clearly to be the right
+choice ergonomically.
+
+## Expansion to arguments
+
+This RFC proposes to allow `impl Trait` in function arguments, in addition to
+return position, with the `any Trait` semantics (as per assumption 4). In other
+words:
+
+```rust
+// These two are equivalent
+fn map<U>(self, f: impl FnOnce(T) -> U) -> Option<U>
+fn map<U, F>(self, f: F) -> Option<U> where F: FnOnce(T) -> U
+```
+
+However, this RFC also proposes to *disallow* use of `impl Trait` within `Fn`
+trait sugar or higher-ranked bounds, i.e. to disallow examples like the following:
+
+```rust
+fn foo(f: impl Fn(impl SomeTrait) -> impl OtherTrait)
+fn bar() -> (impl Fn(impl SomeTrait) -> impl OtherTrait)
+```
+
+While we will eventually want to allow such uses, it's likely that we'll want to
+introduce nested universal quantifications (i.e., higher-ranked bounds) in at
+least some cases; we don't yet have the ability to do so. We can revisit this
+question later on, once higher-ranked bounds have gained full expressiveness.
+
+### Explicit instantiation
+
+This RFC does *not* propose any means of explicitly instantiating an `impl
+Trait` in argument position. In other words:
+
+```rust
+fn foo<T: Trait>(t: T)
+fn bar(t: impl Trait)
+
+foo::<u32>(0) // this is allowed
+bar::<u32>(0) // this is not
+```
+
+Thus, while `impl Trait` in argument position in some sense "desugars" to a
+generic parameter, the parameter is treated fully anonymously.
+
+## Expansion to traits
+
+This RFC proposes to allow `impl Trait` in *function signatures* within both
+trait definitions and trait impl blocks. It does *not* propose allowing `impl
+Trait` to appear in associated type definitions.
+
+Allowing use in traits is motivated by many examples, most prominently:
+
+```rust
+trait Iterable {
+    type Item;
+    fn iter<'a>(&'a self) -> impl Iterator<&'a Item> + 'a
+}
+```
+
+Here `impl Trait` is used for a type whose identity isn't important, where
+introducing an associated type is overkill.
```

Comment:
If I understand correctly, this proposal would only allow returning `impl Trait` when implementing a trait that uses this new syntax. This prevents the use of `impl Trait` return types when implementing trait functions that return an associated type.

This means, for example, that the `Service` trait from Tokio ([link](https://github.com/tokio-rs/tokio-service/blob/master/src/lib.rs#L142)) would have to be changed to the following in order to support `Service`s returning `impl Future`s:

```rust
pub trait Service {

    /// Requests handled by the service.
    type Request;

    /// Responses given by the service.
    type Response;

    /// Errors produced by the service.
    type Error;

    // NOTE: No more `type Future`

    /// Process the request and return the response asynchronously.
    fn call(&self, req: Self::Request) -> impl Future<Item = Self::Response, Error = Self::Error>;
}
```

But now because the `Future` associated type has been removed, it's impossible to place bounds on the output of a `Service`. For example, when writing a  `fn foo<S>(s: S) where S: Service`, it's no longer possible to constrain `<S as Service>::Future: 'static + Sync + Clone + etc.`

Similarly, for the `Iterable` trait provided, it's impossible to write `fn use_exact_size_iterable<I>(x: I) where I: Iterable, <I as Iterable>::Iterator: ExactSizeIterator`.

Because of these restrictions, library authors would be forced to choose whether they want their trait to be usable with `impl Trait`, or whether they want their return types to be boundable. This would be a very difficult and conflicting decision, and either choice would undoubtedly result in frustration for users of the trait who either want to use `impl Trait` or place bounds on return types.